File: src/medium/three-sum/1.go
package main

import (
	"sort"
)

// func main() {
// 	// res := threeSum([]int{-1, 0, 1, 2, -1, -4, -2, -3, 3, 0, 4})
// 	res := threeSum([]int{-4, -2, -2, -2, 0, 1, 2, 2, 2, 3, 3, 4, 4, 6, 6})
// 	// res := threeSum([]int{-1, 0, 1})
// 	fmt.Println("---result----")
// 	fmt.Println(res)
// }

/*
 * @lc app=leetcode id=15 lang=golang
 *
 * [15] 3Sum
 */

// @lc code=start
func threeSum(nums []int) [][]int {
	var result [][]int

	if len(nums) < 3 {
		return result
	}

	sort.Ints(nums)

	for i := 0; i < len(nums)-1; i++ {
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}

		for mid, k := i+1, len(nums)-1; mid < k; {
			if nums[i]+nums[mid]+nums[k] == 0 {
				result = append(result, []int{nums[i], nums[mid], nums[k]})
				mid++
				for mid < k && nums[mid] == nums[mid-1] {
					mid++
				}
			} else if nums[i]+nums[mid]+nums[k] < 0 {
				mid++
			} else if nums[i]+nums[mid]+nums[k] > 0 {
				k--
			}
		}
	}
	return result
}

// @lc code=end

File: src/medium/sort-characters-by-frequency/1.ts
/*
 * @lc app=leetcode id=451 lang=typescript
 *
 * [451] Sort Characters By Frequency
 */

// @lc code=start
function frequencySort(s: string): string {
  let freqency = new Map<string, number>();
  for (const char of s) {
    freqency.set(char, (freqency.get(char) ?? 0) + 1);
  }

  let bucket: string[][] = Array.from({ length: s.length + 1 }, () => []);

  for (const [char, count] of freqency) {
    bucket[count].push(char.repeat(count));
  }

  let result = '';
  for (let i = bucket.length - 1; i >= 0; i--) {
    result += bucket[i].join('');
  }
  return result;
}

// time complexity: O(n)
// space complexity: O(n)
// @lc code=end
